lib/oban/queue/producer.ex

```diff
@@ -256,29 +264,58 @@ defmodule Oban.Queue.Producer do
     {:noreply, state}
   end
 
-  defp dispatch(%State{conf: conf, foreman: foreman} = state) do
-    %State{limit: limit, nonce: nonce, queue: queue, running: running} = state
+  defp dispatch(%State{} = state) do
+    cond do
+      dispatch_now?(state) ->
+        %State{conf: conf, limit: limit, nonce: nonce, queue: queue, running: running} = state
 
-    started_jobs =
-      for job <- fetch_jobs(conf, queue, nonce, limit - map_size(running)), into: %{} do
-        {:ok, pid} = DynamicSupervisor.start_child(foreman, Executor.child_spec(job, conf))
+        running =
+          conf
+          |> fetch_jobs(queue, nonce, limit - map_size(running))
+          |> start_jobs(state)
+          |> Map.merge(running)
 
-        {Process.monitor(pid), {job, pid}}
-      end
+        {:noreply, %{state | cooldown_ref: nil, dispatched_at: unix_now(), running: running}}
+
+      cooldown_available?(state) ->
+        %State{conf: conf, dispatched_at: dispatched_at} = state
+
+        dispatch_after = unix_now() - dispatched_at + conf.dispatch_cooldown
+        cooldown_ref = Process.send_after(self(), :dispatch, dispatch_after)
+
+        {:noreply, %{state | cooldown_ref: cooldown_ref}}
 
-    {:noreply, %{state | running: Map.merge(running, started_jobs)}}
+      true ->
+        {:noreply, state}
+    end
   rescue
     exception in trip_errors() -> {:noreply, trip_circuit(exception, state)}
   end
 
+  defp dispatch_now?(%State{dispatched_at: nil}), do: true
+
+  defp dispatch_now?(%State{conf: conf, dispatched_at: dispatched_at}) do
+    unix_now() > dispatched_at + conf.dispatch_cooldown
+  end
+
+  defp cooldown_available?(%State{cooldown_ref: ref}), do: is_nil(ref)
+
   defp fetch_jobs(conf, queue, nonce, count) do
     case Query.fetch_available_jobs(conf, queue, nonce, count) do
       {0, nil} -> []
       {_count, jobs} -> jobs
     end
   end
 
-  defp running_job_ids(%State{running: running}) do
-    for {_ref, {%_{id: id}, _pid}} <- running, do: id
+  defp start_jobs(jobs, %State{conf: conf, foreman: foreman}) do
+    for job <- jobs, into: %{} do
+      {:ok, pid} = DynamicSupervisor.start_child(foreman, Executor.child_spec(job, conf))
+
+      {Process.monitor(pid), {job, pid}}
+    end
+  end
+
+  defp unix_now do
+    DateTime.to_unix(DateTime.utc_now(), :millisecond)
```

Comment:
It would probably be safer to use monotonic time here: https://hexdocs.pm/elixir/System.html#monotonic_time/1

Reply:
Great point. Changed it up 👍